data/transactions/teal.go

```diff
@@ -0,0 +1,113 @@
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with go-algorand.  If not, see <https://www.gnu.org/licenses/>.
+
+package transactions
+
+import (
+	"github.com/algorand/go-algorand/data/basics"
+)
+
+// LogItem contains logs for an application. ID is the offset into
+// Txn.ForeignApps
+type LogItem struct {
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+
+	ID      uint64 `codec:"i"`
+	Message string `codec:"m"`
+}
+
+// Equal checks whether two LogItems are equal.
+func (l LogItem) Equal(o LogItem) bool {
+
+	return l.ID == o.ID && l.Message == o.Message
+
+}
+
+// EvalDelta stores StateDeltas for an application's global key/value store, as
+// well as StateDeltas for some number of accounts holding local state for that
+// application
+type EvalDelta struct {
+	_struct struct{} `codec:",omitempty,omitemptyarray"`
+
+	GlobalDelta basics.StateDelta `codec:"gd"`
+
+	// When decoding EvalDeltas, the integer key represents an offset into
+	// [txn.Sender, txn.Accounts[0], txn.Accounts[1], ...]
+	LocalDeltas map[uint64]basics.StateDelta `codec:"ld,allocbound=config.MaxEvalDeltaAccounts"`
+
+	Logs []LogItem `codec:"lg,allocbound=config.MaxLogCalls"`
+
+	// Intentionally, temporarily wrong - need to decide how to
+	// allocbound properly when structure is recursive.  Even a bound
+	// of 2 would allow arbitrarily large object if deep.
+	InnerTxns []SignedTxnWithAD `codec:"itx,allocbound=4"`
+}
+
+// Equal compares two EvalDeltas and returns whether or not they are
+// equivalent. It does not care about nilness equality of LocalDeltas,
+// because the msgpack codec will encode/decode an empty map as nil, and we want
+// an empty generated EvalDelta to equal an empty one we decode off the wire.
+func (ed EvalDelta) Equal(o EvalDelta) bool {
+	// LocalDeltas length should be the same
+	if len(ed.LocalDeltas) != len(o.LocalDeltas) {
+		return false
+	}
+
+	// All keys and local StateDeltas should be the same
+	for k, v := range ed.LocalDeltas {
+		// Other LocalDelta must have value for key
+		ov, ok := o.LocalDeltas[k]
+		if !ok {
+			return false
+		}
+
+		// Other LocalDelta must have same value for key
+		if !ov.Equal(v) {
+			return false
+		}
+	}
+
+	// GlobalDeltas must be equal
+	if !ed.GlobalDelta.Equal(o.GlobalDelta) {
+		return false
+	}
+
+	// Logs must be equal
+	if len(ed.Logs) != len(o.Logs) {
+		return false
+	}
+	for i, l := range ed.Logs {
+		ok := l.Equal(o.Logs[i])
+		if !ok {
+			return false
+		}
+	}
+
+	// TODO: check inner transactions are equal
+
+	return true
+}
+
+// SetLogs taks a simple slice of log messages and creates LogItems
+// with a zero ID. LogItem should probably go away, since logs will
+// always appear with the app that produced them, a simple string list
+// would be fine.
```

Comment:
Yes, that's right, which means my whole push to get an app id associated with a log message was a bad idea.  I should rip it out (because now it *must* be the current app).  We do lose the inherent ability to know ordering, callers would have to emit a marker log message when calling apps to ensure they could recover the order if it matters to them. (Not that they can call apps yet, but they will eventually.)

Reply:
LogItem eliminated